convert2rhel/utils.py

```diff
@@ -357,24 +358,37 @@ def install_pkgs(pkgs_to_install, replace=False, critical=True):
 
 
 def download_pkg(pkg, dest=TMP_DIR, disablerepo=None, enablerepo=None):
-    """Download the specified package."""
+    """Download an rpm using yumdownloader and return its path."""
+    loggerinst = logging.getLogger(__name__)
     cmd = "yumdownloader"
     if disablerepo is None:
         disablerepo = []
     if enablerepo is None:
         enablerepo = []
 
     for repo in disablerepo:
-        cmd += " --disablerepo=%s " % repo
+        cmd += " --disablerepo=%s" % repo
 
     for repo in enablerepo:
-        cmd += " --enablerepo=%s " % repo
+        cmd += " --enablerepo=%s" % repo
 
-    cmd += " --destdir %s " % dest
+    cmd += ' --destdir="%s"' % dest
     cmd += " %s" % pkg
 
-    _, ret_code = run_subprocess(cmd, print_output=False)
-    return ret_code
+    output, ret_code = run_subprocess(cmd, print_output=False)
+    if ret_code != 0:
+        loggerinst.warning("Couldn't download the %s package using yumdownloader:\n%s." % (pkg, output))
+        return None
+
+    # The name of the downloaded rpm is on the last line of the output from yumdownloader
+    # The line looks like:
+    #  "vim-enhanced-8.0.1763-13.0.1.el8.x86_64.rpm     2.2 MB/s | 1.4 MB     00:00"
+    rpm_name_match = re.search(r"^.*\.rpm", output.splitlines()[-1])
```

Comment:
Breaks on OL6 
```py
[10/30/2020 11:05:48] DEBUG - Calling command 'yumdownloader --destdir="/var/lib/convert2rhel/" 1:redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64'
[10/30/2020 11:05:49] DEBUG - Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/convert2rhel/main.py", line 80, in main
    pre_ponr_conversion()
  File "/usr/lib/python2.6/site-packages/convert2rhel/main.py", line 154, in pre_ponr_conversion
    pkghandler.remove_excluded_pkgs()
  File "/usr/lib/python2.6/site-packages/convert2rhel/pkghandler.py", line 385, in remove_excluded_pkgs
    utils.remove_pkgs([get_pkg_nevra(pkg) for pkg in installed_excluded_pkgs])
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 308, in remove_pkgs
    changed_pkgs_control.backup_and_track_removed_pkg(nvra)
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 254, in backup_and_track_removed_pkg
    restorable_pkg.backup()
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 457, in backup
    self.path = download_pkg(self.name)
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 386, in download_pkg
    rpm_name_match = re.search(r"^.*\.rpm", output.splitlines()[-1])
IndexError: list index out of range

WARNING - Abnormal exit! Performing rollback ...
```
And following that is a second error, but seems unrelated to this PR 
```
[10/30/2020 11:09:47] TASK - [Rollback: Installing removed packages] ****************************
Installing packages:
        /var/lib/convert2rhel/redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64.rpm
        /var/lib/convert2rhel/oraclelinux-release-6Server-10.0.2.x86_64.rpm
        /var/lib/convert2rhel/oraclelinux-release-notes-6Server-22.0.2.el6.x86_64.rpm
[10/30/2020 11:09:47] DEBUG - Calling command 'rpm -i /var/lib/convert2rhel/redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64.rpm /var/lib/convert2rhel/oraclelinux-release-6Server-10.0.2.x86_64.rpm /var/lib/convert2rhel/oraclelinux-release-notes-6Server-22.0.2.el6.x86_64.rpm'
Traceback (most recent call last):
  File "/usr/bin/convert2rhel", line 9, in <module>
    load_entry_point('convert2rhel==0.12', 'console_scripts', 'convert2rhel')()
  File "/usr/lib/python2.6/site-packages/convert2rhel/main.py", line 116, in main
    rollback_changes()
  File "/usr/lib/python2.6/site-packages/convert2rhel/main.py", line 219, in rollback_changes
    utils.changed_pkgs_control.restore_pkgs()
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 281, in restore_pkgs
    self._install_removed_pkgs()
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 275, in _install_removed_pkgs
    install_pkgs(pkgs_to_install, critical=False)
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 343, in install_pkgs
    output, ret_code = run_subprocess("%s %s" % (cmd, pkgs), print_output=False)
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 135, in run_subprocess
    env={'LC_ALL':'C'})
  File "/usr/lib64/python2.6/subprocess.py", line 642, in __init__
    errread, errwrite)
  File "/usr/lib64/python2.6/subprocess.py", line 1238, in _execute_child
    raise child_exception
TypeError: execve() argument 1 must be encoded string without NULL bytes, not str
```

Reply:
This is from spinning up our Vagrantbox using `PR=112 SYSTEM=ol6 vagrant up`

Ran the exact same code as you:
```
# rpm -qa convert2rhel
convert2rhel-0.12-1.20201030155418021773.pr112.33.g92ee17e.el6.noarch
# rpm -qa yum-utils
yum-utils-1.1.30-42.0.1.el6_10.noarch
# yumdownloader --destdir="/var/lib/convert2rhel/" bogus
Failed to set locale, defaulting to C
No Match for argument bogus
Nothing to download
# echo $?
1
# yumdownloader --destdir="/var/lib/convert2rhel/" 1:redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64
Failed to set locale, defaulting to C
redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64.rpm                                  | 2.6 kB     00:00     
# python
>>> from convert2rhel import utils
>>> utils.download_pkg("1:redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64")
u'/var/lib/convert2rhel/redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64.rpm'
```
Then I ran `convert2rhel --debug -y --disable-submgr --enablerepo rhel-6-server-rpms --enablerepo rhel-6-server-optional-rpms`

Reply:
Interestingly enough, there's a problem with yumdownloader. Its output goes neither to stdout nor stderr :-o I'm not sure what to do about that :( I don't know where it goes then.
```
# rm -rf /var/lib/convert2rhel/*
# yumdownloader --destdir='/var/lib/convert2rhel/' 1:redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64 >out 2>&1
# echo $?
0
# ls /var/lib/convert2rhel/
redhat-release-server-6Server-6.10.0.12.0.1.el6.x86_64.rpm
# cat out
# 
```

Reply:
So, I haven't gone through all the historical code of yum available on OL 6 but it most certainly yumdownloader does not print anything out when the stdout is not an active terminal, like the condition here: https://github.com/rpm-software-management/yum-utils/blob/master/repoquery.py#L1584.
There's not much to do about that. It seems there's no way to get the path of the downloaded rpm out of yumdownloader. I was thinking about using `yum install --downloadonly` but that checks if the package is already installed and if so, it refuses to download it.
I'll probably workaround it in a way it was there before - download the rpm with yumdownloader, not check its output and then try to find it in the destination folder :'(

Reply:
Inspired by https://terminallabs.com/blog/a-better-cli-passthrough-in-python/, I've used pexpect: https://github.com/oamg/convert2rhel/pull/112/commits/26d39ba7b924b0fb034808c13d036fde1086b5f0

Still, this is not working on CentOS/OL 7. I will fix it later.

Reply:
Works on CentOS/OL 7 as well now.